luarules/gadgets/unit_area_timed_damage.lua

```diff
@@ -16,225 +16,521 @@ if not gadgetHandler:IsSyncedCode() then
+				startTimedArea(queue[ii], queue[ii+1], queue[ii+2], queue[ii+3], queue[ii+4])
+			end
+		end
+		delayQueue[gameFrame] = nil
+	end
 end
 
-function gadget:UnitDestroyed(unitID, unitDefID, unitTeam, attackerID, attackerDefID, attackerTeam)
-	local params = OnDeathAreaDefs[unitDefID]
-	local ux, uy, uz = Spring.GetUnitPosition(unitID)
-	StartTimedArea(ux, uy, uz, params)
+function gadget:Explosion(weaponDefID, px, py, pz, attackID, projID)
+	if weaponTriggerParams[weaponDefID] then
+		startTimedArea(px, py, pz, weaponTriggerParams[weaponDefID], (attackID or -1))
+	end
 end
 
-function gadget:GameFrame(frame)
-	UpdateTimedAreas(frame)
+function gadget:UnitDestroyed(unitID, unitDefID, unitTeam, attackID, attackDefID, attackTeam)
+	if destroyTriggerParams[unitDefID] then
+		local ux, uy, uz = spGetUnitPosition(unitID)
+		startTimedArea(ux, uy, uz, destroyTriggerParams[unitDefID], -1)
+	end
+	shieldUnits[unitID] = nil
 end
 
 function gadget:UnitPreDamaged(unitID, unitDefID, unitTeam, damage, paralyzer, weaponDefID, projID, attackID, attackDefID, attackTeam)
-	local params = TimedAreaDefs[weaponDefID]
-	if params then
-		local unitDef = UnitDefs[unitDefID]
-		local immunity = unitDef.customParams.areadamageresistance
-		if immunity and string.find(immunity, params.atd_resistance, 1, true) then
+	if timedAreaParams[weaponDefID] then
+		local immunities = damageImmunities[unitDefID]
+		if immunities and (immunities[timedAreaParams[weaponDefID].area_damagetype] or immunities.all) then
 			return 0, 0
-		elseif params.atd_damage_ceg then
-			local _,_,_, x,y,z = Spring.GetUnitPosition(unitID, true)
-			if y > 0 then
-				Spring.SpawnCEG(params.atd_damage_ceg, x, y + 8, z, 0, 0, 0, 0, damage)
-			else
-				return 0, 0
-			end
-			-- return damage * damageRate, impulse * impulseRate
 		end
 	end
-	return damage, 1
 end
 
-function gadget:FeaturePreDamaged(featureID, featureTeam, damage, weaponDefID, projID, attackID, attackDefID, attackTeam)
-	local params = TimedAreaDefs[weaponDefID]
-	if params and params.atd_damage_ceg then
-		local _,_,_, x,y,z = Spring.GetFeaturePosition(featureID, true)
-		Spring.SpawnCEG(params.atd_damage_ceg, x, y + 8, z, 0, 0, 0, 0, damage)
+function gadget:FeaturePreDamaged(featureID, featureDefID, featureTeam, damage, weaponDefID, projID, attackID, attackDefID, attackTeam)
+	if timedAreaParams[weaponDefID] then
+		local damagedCEG = timedAreaParams[weaponDefID].area_damagedceg
+		if not ignoredFeatureDefs[featureDefID] and damagedCEG then
+			local _,_,_, x,y,z = spGetFeaturePosition(featureID, true)
+			spSpawnCEG(damagedCEG, x, y + 18, z)
+		end
+	end
+end
+
+function gadget:UnitDamaged(unitID, unitDefID, unitTeam, damage, paralyzer, weaponDefID, projID, attackID, attackDefID, attackTeam)
+	if timedAreaParams[weaponDefID] then
+		local damagedCEG = timedAreaParams[weaponDefID].area_damagedceg
+		if damagedCEG then
+			local _,_,_, x,y,z = spGetUnitPosition(unitID, true)
+			spSpawnCEG(damagedCEG, x, y + 18, z)
```

Comment:
Why `y + 18`? What is this magic number?

Reply:
A nice version of this might do something like `y + unitRadius[unitDefID]`. This is a cheap offset to try to make sure the CEG shows. I think it was y+8 before.